src/compiler/checker.ts

```diff
@@ -2277,8 +2278,8 @@ namespace ts {
             result.declarations = deduplicate(concatenate(valueSymbol.declarations, typeSymbol.declarations), equateValues);
             result.parent = valueSymbol.parent || typeSymbol.parent;
             if (valueSymbol.valueDeclaration) result.valueDeclaration = valueSymbol.valueDeclaration;
-            if (typeSymbol.members) result.members = typeSymbol.members;
-            if (valueSymbol.exports) result.exports = valueSymbol.exports;
+            if (typeSymbol.members) result.members = cloneMap(typeSymbol.members);
+            if (valueSymbol.exports) result.exports = cloneMap(valueSymbol.exports);

```

Comment:
Why do we need to clone here?

Reply:
if `typeSymbol` or `valueSymbol` are _not_ `TransietSymbol`s, then their symbol tables consist only of non-transient symbols and persist through multiple checks. If we only copy the reference to the table into a new `TransientSymbol`, when we later stuff more things into the table (as is possible for transient symbols), we'll pollute the supposedly "immutable" symbol table from the original non-transient symbol with new transient members.

Reply:
This was one of 3 possible leaks. We assume elsewhere in the checker that any `SymbolTable` attached to a `TransientSymbol` is safe to mutate, and sometimes add new symbols to it. `combineValueAndTypeSymbols` currently just blindly assigns the reference to the symbol table to `result` (which is a `TransientSymbol`), ignoring whether `typeSymbol` or `valueSymbol` were created by the binder. As a result, there is the possibility of some other function in checker later mutating the `SymbolTable`.